### ISSUE TITLE
Update exercises based on "Teaching Rust to Python developers"

### ISSUE THREAD

**Issue by djmitche** [google/comprehensive-rust#2525]
See #2502 and in particular, from @randomPoison:
> Oh that's great. As much as I like the protobuf exercise, it is pretty complicated and I think students spend a lot of time just wrapping their heads around what they need to do. This seems way more concise and understandable, and better demonstrates the usage of lifetimes.
> 
> I think both of these are nice improvements over our current ones, so I think these would be great to integrate :)

**Comment by djmitche** [comment]
I don't think this is actionable by anyone but me, and it appears I'm not going to do it :)

**Comment by mgeisler** [comment]
I think we need to reopen this: I am teaching Rust Fundamentals right now and we reached this exercise today. Even with me spending 10-15 minutes on

- Briefly explaining what protocol buffer are,
- Explaining the idea of parsing a slice of bytes that shrink in size
- Explaining zero-copy parsing

people still spent 15 minutes getting themselves oriented in the code. I think some could have solved this in 45 minutes, perhaps, but most cannot.

The comment from the participants in my class is simple that the exercise has too much emphasis on non-Rust concepts.

**Comment by mgeisler** [comment]
@gribozavr, would you be able to find someone at Google who can look at reworking this?

**Comment by marc-hb** [comment]
I just did the protobuf exercise on my own; no class and no teacher. I didn't find it too bad in itself. I think it was decent rust practice and parsing is a memory safety "poster child"! But I didn't find that exercise taught me much about lifetimes despite being located there right now. 

> people still spent 15 minutes getting themselves oriented in the code. I think some could have solved this in 45 minutes, perhaps, but most cannot.
> The comment from the participants in my class is simple that the exercise has too much emphasis on non-Rust concepts.

Agreed! It did take me a significant, "non-rust-specific" time to educate myself about protobufs and orient myself in the code and to find what was missing, so maybe not the "best value".

I found this bit of information was especially missing, I suspect it could have saved me a fair amount of time:

```diff
--- a/src/lifetimes/exercise.rs
+++ b/src/lifetimes/exercise.rs
@@ -47,6 +47,10 @@ struct Field<'a> {
     value: FieldValue<'a>,
 }
 
+// Parsing happens in TWO STEPS:
+// 1. bytes -> Protobuf fields,
+// 2. Protobuf fields -> Person / PhoneNumber.
+// "add_field()" performs the second step
 trait ProtoMessage<'a>: Default {
     fn add_field(&mut self, field: Field<'a>);
 }
```
(Either there or in any better place)

PS: the course is great, thank you!
PPS: not sure what was @djmitche's initial intention but having the "protobuf" keyword in the issue title would be useful now.

**Comment by djmitche** [comment]
>  PPS: not sure what was @djmitche's initial intention but having the "protobuf" keyword in the issue title would be useful now.

Good point, I've changed the title

**Comment by mgeisler** [comment]
> I think it was decent rust practice and parsing is a memory safety "poster child"!

Good point and I do find the exercise very cool from that perspective!
